### ISSUE TITLE
TFT Model not reproducible

### ISSUE THREAD

**Issue by sonja-kompzkfe** [unit8co/darts#2033]
I am trying to train a TFT model but have issues with reproducible results. If I create two tft models with QuantileRegression likelihoods with identical parameters (and same random state) and train & predict, I get different results. I set all seeds as recommended by pytorch (https://pytorch.org/docs/stable/notes/randomness.html) and the Darts  TFT tutorial. The longer I train the models (more epochs), the more profound the differences are. However if I make deterministic predictions with the MSEloss, the results are almost identical for the two models (approx. differences in the 6th position after the decimal point). Although I work with a custom dataset, I also encountered this problem with the AirPassenger dataset following the tft tutorial.

Also, I am wondering if it is possible to get reproducible results in the following scenario: I am training a model and call predict() one time. If I then rerun my script (training the model and calling predict one time) should the predictions be identical? In my case, they are different.

My code is from the TFT tutorial with the exception that I add a second model:

```
import numpy as np
import pandas as pd
from tqdm import tqdm_notebook as tqdm

import matplotlib.pyplot as plt

from darts import TimeSeries
from darts.dataprocessing.transformers import Scaler
from darts.models import TFTModel
from darts.datasets import AirPassengersDataset
from darts.utils.timeseries_generation import datetime_attribute_timeseries
from darts.utils.likelihood_models import QuantileRegression

import warnings
import random
import os
import torch
import lightning.pytorch as pl
seed_value =1
random.seed(seed_value)
np.random.seed(seed_value)
torch.manual_seed(seed_value)
torch.cuda.manual_seed(seed_value)
torch.cuda.manual_seed_all(seed_value)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
pl.seed_everything(seed_value, workers=True)

os.environ["CUBLAS_WORKSPACE_CONFIG"]=":4096:2"
os.environ["CUBLAS_WORKSPACE_CONFIG"]=":16:8"
os.environ["CUDA_LAUNCH_BLOCKING"] = "1"
torch.use_deterministic_algorithms(True, warn_only=True)

warnings.filterwarnings("ignore")
import logging

logging.disable(logging.CRITICAL)
# Read data
series = AirPassengersDataset().load()

# we convert monthly number of passengers to average daily number of passengers per month
series = series / TimeSeries.from_series(series.time_index.days_in_month)
series = series.astype(np.float32)

# Create training and validation sets:
training_cutoff = pd.Timestamp("19571201")
train, val = series.split_after(training_cutoff)

# Normalize the time series (note: we avoid fitting the transformer on the validation set)
transformer = Scaler()
train_transformed = transformer.fit_transform(train)
val_transformed = transformer.transform(val)
series_transformed = transformer.transform(series)

# create year, month and integer index covariate series
covariates = datetime_attribute_timeseries(series, attribute="year", one_hot=False)
covariates = covariates.stack(
    datetime_attribute_timeseries(series, attribute="month", one_hot=False)
)
covariates = covariates.stack(
    TimeSeries.from_times_and_values(
        times=series.time_index,
        values=np.arange(len(series)),
        columns=["linear_increase"],
    )
)
covariates = covariates.astype(np.float32)

# transform covariates (note: we fit the transformer on train split and can then transform the entire covariates series)
scaler_covs = Scaler()
cov_train, cov_val = covariates.split_after(training_cutoff)
scaler_covs.fit(cov_train)
covariates_transformed = scaler_covs.transform(covariates)

# default quantiles for QuantileRegression
quantiles = [0.01,0.05,0.1,0.15,0.2,0.25,0.3,0.4,0.5,0.6,0.7,0.75,0.8,0.85,0.9,0.95,0.99,]
input_chunk_length = 24
forecast_horizon = 12
my_model = TFTModel(
    input_chunk_length=input_chunk_length,
    output_chunk_length=forecast_horizon,
    hidden_size=64,
    lstm_layers=1,
    num_attention_heads=4,
    dropout=0.1,
    batch_size=16,
    n_epochs=300,
    add_relative_index=False,
    add_encoders=None,
    likelihood=QuantileRegression(
        quantiles=quantiles
    ),  # QuantileRegression is set per default
    #loss_fn=torch.nn.MSELoss(),
    random_state=0,
)

my_model1 = TFTModel(
    input_chunk_length=input_chunk_length,
    output_chunk_length=forecast_horizon,
    hidden_size=64,
    lstm_layers=1,
    num_attention_heads=4,
    dropout=0.1,
    batch_size=16,
    n_epochs=300,
    add_relative_index=False,
    add_encoders=None,
    likelihood=QuantileRegression(
        quantiles=quantiles
    ),  # QuantileRegression is set per default
    #loss_fn=torch.nn.MSELoss(),
    random_state=0,
)
my_model.fit(train_transformed, future_covariates=covariates_transformed, verbose=True)
my_model1.fit(train_transformed, future_covariates=covariates_transformed, verbose=True)

pred_series = my_model.predict(n=1, num_samples=10)
pred_series1 = my_model1.predict(n=1, num_samples=10)`

```
Am I missing something obvious here? I use darts==0.26.0 and python 3.10.

**Comment by madtoinou** [comment]
Hi @sonja-kompzkfe,

I ran your code snippet on my Apple Silicon laptop and the forecasts are exactly identical after training the models for either 20 or 200 epochs. This would indicate that the slight deviations that you are reporting might be caused by the GPU/cuda. Can you confirm that you are using GPU? If no, which OS are you using?

**Comment by sonja-kompzkfe** [comment]
Hi @madtoinou,

Thanks for looking into this. Yes I am using a GPU (NVIDIA Quadro T1000).

**Comment by dennisbader** [comment]
Hi @sonja-kompzkfe, reproducibility should already be handled on the Darts and PyTorch Lightning side, so no need to manually seed.

Can you try after removing all the related lines up and including `torch.use_deterministic_algorithms`?

**Comment by sonja-kompzkfe** [comment]
@dennisbader, sorry for my late reply. I did remove all the lines you suggested and the results still are not reproducible. I also set up a new clean environment and it didn't help (I am working on a windows machine). If I run the code in a Google Colab notebook, the results are reproducible, so I assume it has to be an issue with my GPU. If I don't use a GPU on my local machine, results are reproducible.
